actions/concurrent_runner.go

```diff
@@ -0,0 +1,113 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package actions
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const defaultTimeout = time.Second * 10
+
+// ActionResult represents an action result.
+//nolint:unused
+type ActionResult struct {
+	ID             string
+	Type           string
+	Error          error
+	CombinedOutput []byte
+}
+
+// ConcurrentRunner represents concurrent action runner.
+// Action runner is component that can run an Actions.
+//nolint:unused
+type ConcurrentRunner struct {
+	out chan ActionResult
+	l   *logrus.Entry
+
+	// runningActions stores CancelFunc's for running actions.
+	runningActions sync.Map // map[string]CancelFunc
```

Comment:
When we are talking about a few actions, the fastest data structure is actually array/slice, not map or sync.Map. The latter is used when both the number of elements and number of CPU cores accessing them. That's hardly our case.
But we also don't care about performance in this case. What we care about is type safety. `interface{}` is not type safe.

Reply:
I don't understand what are you talking about? About premature optimization or about what? If we are talking about speed, linear search in HashMap is `O(1)` and on the array is `O(N)`. What the problem? 

The latter is used in 2 cases:

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or **(2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys.** In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

Why this code is not the 2nd case?